lib/Sema/TypeCheckProtocol.cpp

```diff
@@ -1255,9 +1252,16 @@ matchWitness(TypeChecker &tc,
   };
 
   // Match a type in the requirement to a type in the witness.
-  auto matchTypes = [&](Type reqType, Type witnessType) 
+  auto matchTypes = [&](Type reqType, Type witnessType)
                       -> Optional<RequirementMatch> {
-    cs->addConstraint(ConstraintKind::Equal, reqType, witnessType, locator);
+    ConstraintKind kind = ConstraintKind::Conversion;
+    if (auto func = dyn_cast<FuncDecl>(req)) {
+      auto name = req->getFullName().getBaseName();
+      if (func->isOperator() && name.str() == "==") {
+        kind = ConstraintKind::Equal;
+      }
+    }
+    cs->addConstraint(kind, witnessType, reqType, locator);
```

Comment:
Oh. So I think the problem is for function arguments, we need to match the witness against the requirement; for results, we need to match the requirement against the witness. This should give you the correct variance and eliminate the need to swap the arguments in matchTypes() above.